### PR TITLE
[rocdl] Adjust heuristic seeds for smaller/large matmuls

### PR DESCRIPTION
**Pull request by Groverkss**
For smaller matmuls we want to distribute to more
workgroups in order to fill the GPU better. So use
a smaller tile sizes for M and N.

For larger matmuls we can also use a larger K tile
size to issue more mfma instructions to fill compute
units.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 1]
```diff
@@ -374,12 +374,27 @@ setConvolutionVectorDistributionConfig(mlir::FunctionOpInterface entryPoint,
...
-  // See https://github.com/openxla/iree/issues/16341 for details.
-  GPUMMAHeuristicSeeds seeds{/*bestSubgroupCountPerWorkgroup=*/4,
-                             /*bestMNTileCountPerSubgroup=*/8,
-                             /*bestKTileCountPerSubgroup=*/2};
+  GPUMMAHeuristicSeeds seeds;
+
+  int64_t mnSize = problem.mSize * problem.nSize;
+  // TODO: We should calculate this value based on the target. This
+  // is also highly dependent on the intrinsic we are targeting.
+  int64_t mnSizeThreshold = 163840;
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: Can you explain where this initial value came from?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 2]
```diff
@@ -374,12 +374,27 @@ setConvolutionVectorDistributionConfig(mlir::FunctionOpInterface entryPoint,
...
+  // workgroups by using a smaller bestMNTileCountPerSubgroup and larger
+  // bestKTileCountPerSubgroup.
+  if (mnSize <= mnSizeThreshold) {
+    seeds = {/*bestSubgroupCountPerWorkgroup=*/4,
+             /*bestMNTileCountPerSubgroup=*/4,
+             /*bestKTileCountPerSubgroup=*/8};
+  } else {
+    // Note that the following heuristic seeds are just placeholder values.
+    // We need to clean it up and make it adjusting to different targets.
+    // See https://github.com/openxla/iree/issues/16341 for details.
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: Move this comment to the top, and put the comment about small M * N inside the if-then block.

**Review by qedawkins — CHANGES_REQUESTED** [review]
Please add a test. Also it's nice to add descriptions to PRs like this giving some motivation behind the change (similar to what you have in the comment).

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp (hunk 1 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: +1.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 3]
```diff
@@ -374,12 +374,27 @@ setConvolutionVectorDistributionConfig(mlir::FunctionOpInterface entryPoint,
     intrinsics.emplace_back(mSize, nSize, kSize, aType, bType, cType);
   }
 
-  // Note that the following heuristic seeds are just placeholder values.
-  // We need to clean it up and make it adjusting to different targets.
-  // See https://github.com/openxla/iree/issues/16341 for details.
-  GPUMMAHeuristicSeeds seeds{/*bestSubgroupCountPerWorkgroup=*/4,
-                             /*bestMNTileCountPerSubgroup=*/8,
-                             /*bestKTileCountPerSubgroup=*/2};
+  GPUMMAHeuristicSeeds seeds;
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: We need to make the change on the matmul configuration. This is inside the convolution one.

**Review by qedawkins — APPROVED** [review]

**Pull request merged**